add new offer layout and css

KochTobi opened the pull request:
This PR changes the way offers are created.
The new template uses CSS to achieve page breakage.
New offer items are inserted using rows in the corresponding tables which are selected by their CSS queries.
[2021_06_15_O_greiner_sggc_1.pdf](https://github.com/qbicsoftware/offer-manager-2-portlet/files/7114590/2021_06_15_O_greiner_sggc_1.pdf)

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -53,10 +51,12 @@ class OfferToPDFConverter implements OfferExporter {
-                    .getResource("offer-template/offer.html")
-                    .toURI())
+            Paths.get(OFFER_HTML_TEMPLATE_URI)
     private static final Path OFFER_HEADER_IMAGE =
             Paths.get(OfferToPDFConverter.class.getClassLoader()
                     .getResource("offer-template/offer_header.png")
```
Why not also adjusting it here as you have done above? I think it would make sense to fetch the uri first and then get the path either for all of the documents or not to do it like this. (have a consistent way, at least in one class)

KochTobi replied:
In this case, I need the URI to create a static template. For the resource files, it does not make sense to separately construct the URI right now. I removed the unused `Offer_HTML_TEMPLATE` constant.

jenniferboedker commented on offer-manager-domain/src/main/groovy/life/qbic/business/offers/Currency.groovy:
```diff
@@ -11,15 +11,22 @@ import java.text.NumberFormat
 class Currency {
 
-    private final static NumberFormat numberFormat = NumberFormat.getInstance(Locale.US)
+    final static String SYMBOL = "€"
+    private final static NumberFormat NUMBER_FORMAT = NumberFormat.getInstance(Locale.US)
+
+    static String format(double value) {
+        DecimalFormat decimalFormat = getFormatterWithoutSymbol()
+        return decimalFormat.format(value)
+    }
```
Why did you need to change that here? It worked before on my machine. Was it an issue at your side or is it just a "cosmetic" change (if so why)?

KochTobi replied:
By delegating the decision on the symbol away from here we can change the currency more flexible. The main point was however that this forces all currencies to be formatted equally. Previously, a change in the formatterWithoutCurrency was not reflected in the formatterWithCurrency.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+     * Creates a new OfferTemplate using the provided document as template document.
+     * <p>Please note: It is important that the template document provides the following css selectors</p>
+     * <ul>
+     *  <li>.currency-symbol</li>
```
why do we need the currency symbol here? We agreed that since its a german offer that there will be no other currency than €

KochTobi replied:
I was not aware of this decision. This way the offer will only hold € 🤷 since it is set here explicitly

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+            element.text(Currency.format(offer.getOverheadsDataGeneration()))
+        }
+        document.select(".overheads  .data-analysis  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsDataAnalysis()))
```
Its misleading that we have a currency class that does formatting of the number but does not provide a currency per soe

KochTobi replied:
It does provide the currency with its constant field `SYMBOL`. The currency value formatting is done by the currency. The same could have been said about the previous solution with added complexity.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        // fill the table
+        int position = 1
+        Element tableBody = dataGenerationTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataGenerationItems()) {
```
or you do offer.getDataGenerationItems().each{ ... }

KochTobi replied:
the for loop is faster and needs less resources

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        int position = 1
+        Element tableBody = dataGenerationTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataGenerationItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
```
Why come up with a new name again. I fear we will lose track of what is meant when. Why not call it OfferItem. And in general, do we really need to create that object? 
I would like it more if you would call a mehtod on you class that obtains the information from the postition and the offerItem itself. I dont see why you would need that object. The same for the discount. It makes is more complex than it needs to be

KochTobi replied:
I disagree. The OfferPosition is the position in HTML. It does not provide all the information and there might be multiple positions (e.g. discounts, ...) for one offer item. 
The class delegates the creation of the actual HTML body for a row in the listing.

jenniferboedker replied:
But then its a discount item and a normal offer item. I still dont think its a good idea to introduce yet another name which just has the same information as the other DTOs. I think the delegation for the HTML creation is fine, I dont get why you would add a second step to create an object from an object which in essence has the same information

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
```
See my comment above. Furthermore, do you really need to provide position -1. Since we agreed on displaying the discount after the item always I think this can be also achieved with only providing the position itself

KochTobi replied:
It is implemented this way because the logic on where the discount is listed is decided by the template, not the row itself. The OfferPosition is only responsible for providing the HTML for a discount for a selected position on the offer.

jenniferboedker replied:
ok.. but then, instead of providing two positions, which is really confusing to read, how about providing the discount descriptions as string? This is why you need the second position, right? And then the offer template determines how the text looks like.

KochTobi replied:
Sadly this does not work. The description has to span multiple columns so we need two rows here.

KochTobi replied:
You can look at it this way: The OfferItem is the DTO and the OfferPosition the logical entity

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        // fill the table
+        int position = 1
+        Element tableBody = dataAnalysisTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataAnalysisItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
+                tableBody.append(discountPosition.outerHtml())
+                position++
+            }
+        }
```
This is repeated for each table. I think you could outsource that to a method and call it for each of the product groups

KochTobi replied:
You have got a point here :+1: I will do that.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        }
+        // set the footer
+        String netCosts = Currency.format(offer.getNetPMandDS())
+        document.select("#data-management-items > tfoot .costs").first().text(netCosts)
```
why is the css style here different than for the other tables?

KochTobi replied:
good catch.
```suggestion
        dataManagementTable.select(" > tfoot .costs").first().text(netCosts)
```

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        private final double total
+        private final String unit
+
+        static OfferPosition createProductItem(int position, OfferItem offerItem) {
+            String name = offerItem.getProductName()
+            String description = createProductItemDescription(offerItem.getProductDescription(), offerItem.getServiceProvider())
+            double quantity = offerItem.getQuantity()
+            double unitPrice = offerItem.getUnitPrice()
+            double total = offerItem.getItemTotal()
+            String unit = offerItem.getUnit()
+            return new OfferPosition(position, name, description, quantity, unitPrice, total, unit)
+        }
```
see above, I dont think we need that

KochTobi replied:
I think we do. We would have to move the `outerHtml` code here if we did not have the instance with the fields set.
The `createProductItem` or `createDiscountItem` methods would have a String return value. This would not allow us to decouple the creation and the printing of the information.

jenniferboedker replied:
But we decided to not distinguish between offer item and discount. I see now what you have intended to do, but I think this is not necessary. We create the HTML code here and just want to set the fields. I would also argue we "created" the information for the HTML already when we created the offer content.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        document.select("#data-management-items > tfoot .costs").first().text(netCosts)
+    }
+
+    private static class OfferPosition {
```
Naming! See above

```suggestion
    private static class OfferItemGenerator {
```

KochTobi replied:
This would force us to return the string on the createDiscount and createProductItem methods. I do not like that

jenniferboedker replied:
And why is that bad?

KochTobi replied:
because we would have to use it on the spot instead of combining it here.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+            this.unit = unit
+        }
+
+        /**
+         * Generates a html representation of this offer position
+         * @return a html string representing this position
+         * @since 1.1.0
+         */
+        String outerHtml() {
+            String quantity = this.quantity.toString()
+            String unitPrice = Currency.format(this.unitPrice)
+            String total = Currency.format(this.total)
+            String html = """\
+            <tr>
+              <td class="item-number-column">${position}</td>
+              <td class="item-description-column">
+                <div class="item-title">${name}</div>
+              </td>
+              <td class="item-amount-column">${quantity}</td>
+              <td class="item-unit-column">${unit}</td>
+              <td class="item-unit-price-column">${unitPrice}</td>
+              <td class="item-total-column">${total}</td>
+            </tr>
+            <tr>
+              <td class="item-number-column"></td>
+              <td class="item-description-column" colspan="3">
+                <div class="item-description">
+                  ${description}
+                </div>
+              </td>
+            </tr>\
+            """.stripIndent()
+            return html
+        }
```
I like that! Looks very clean and easy to read!

KochTobi replied:
This clarity can only be achieved because we have the fields holding relevant information.
This is why I introduced the OfferPosition

jenniferboedker reviewed (dismissed):
Thanks for all the work you put into this! It will make all our lives easier in the future 😊 

I added some questions for better understanding and improvement of the already really nice code you provided.
Some comments to the final PDF: is it possible to remove small lines for the cost summary?
They are grey in your offer for now. I would argue that it should look like the original offer. Also the long line above overhead total moved below the total. Also I find the spacing between the lines of the cost summary (the final one too small, it looks "pressured".
Other than that, I like that you aligned the addresses for the ncct members in the header :+1:

KochTobi commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        return result
+    }
+
+    private static void fillGeneralInformation(Document document, OfferContent offer) {
```
> why is `document` provided through the method? Why not use the class variable you introduced? And make the method non-static
> 
> the same for all the other methods

This is because we generate a new document from the template and don't modify the template itself.

KochTobi replied:
> Maybe add a hint so that is clear that this method should be run at the end after all information is filled? Because you add the currency then. But I also dont like that it's done like this. Why not handling that with a Formatter and forward the price already with the currency?

Because that adds complexity we do not need. And because we hard-coded the currency symbol in the HTML and css before. This is now no longer the case and is handled by the business object Currency.

KochTobi replied:
> fill general might not really describes what you do here, its more like fillSpecialCharacter.

We do not only set special characters here. What we do is, we set offer global variables like overhead percentage, vat percentage and currency symbol

jenniferboedker replied:
But if you introduce a class variables for the Document and OfferContent (and overwrite them each time fill() is called) you would improve readability of the code and make the methods easier to read. I think the parameters are not necessary here. Its a CleanCode thing to try to reduce the parameters and improve the readability

KochTobi replied:
It is very important to reduce side effects and the readability is not decreased by this I think.

jenniferboedker commented:
When clicking on the QuotationDetails link on the offer it jumps to the datageneration table. I would expect it to just jump to the page. Was it like that before?

KochTobi commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+
+    private static void fillCustomerInformation(Document document, OfferContent offer) {
+        String customerName = String.format(
+                "%s %s %s",
+                offer.getCustomerTitle(),
```
I copied the previous solution for that here. I would suspect it works the same?

KochTobi replied:
has to be tested

KochTobi replied:
No it does not 👍

KochTobi commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    private static void fillProjectManagerInformation(Document document, OfferContent offer) {
+        String projectManagerName = String.format(
+                "%s %s %s",
```
has to be tested.

KochTobi replied:
No it does not :+1:

KochTobi commented:
> Thanks for all the work you put into this! It will make all our lives easier in the future 😊
> 
> I added some questions for better understanding and improvement of the already really nice code you provided.

Thanks a lot for the detailed review! You made some very good suggestions and I implemented some of them. 
> Some comments to the final PDF: is it possible to remove small lines for the cost summary?
> They are grey in your offer for now. I would argue that it should look like the original offer. 

Yes, this is very much possible. I included them as a preparation to allow for cost sections (e.g. for the overheads you could specify the different categories here directly and it would not be separated by a line). I can remove them for now if this is the requested behaviour. 
> Also the long line above overhead total moved below the total. 

Yes, this is because the overheads are specified in the cost summary again. The first one is only a sum of the overheads and is listed again in the actual cost summary. This lead to confusion in the past and I was very confused by this, as well.
> Also I find the spacing between the lines of the cost summary (the final one too small, it looks "pressured".

This can be changed very easily. I did not include additional white space here since this was a request often heard in the past. We can add whitespace if required. (I think it looks nice to have the cost summary values near to each other so that you can recalculate them easier on paper if need be 😉 )

> Other than that, I like that you aligned the addresses for the ncct members in the header 👍

Oh, I did? Sady, I think this only happened by chance 😄 .

jenniferboedker commented:
> Yes, this is very much possible. I included them as a preparation to allow for cost sections (e.g. for the overheads you could specify the different categories here directly and it would not be separated by a line). I can remove them for now if this is the requested behaviour.

Yes, please remove them.

> Yes, this is because the overheads are specified in the cost summary again. The first one is only a sum of the overheads and is listed again in the actual cost summary. This lead to confusion in the past and I was very confused by this, as well.
I understand that and think its very cool, that you came up with new ideas but would kindly also ask you to put the line where it was. This is something the stakeholders need to request.

Please try to stick to what was already approved:
<img width="921" alt="Screenshot 2021-08-19 at 09 01 22" src="https://user-images.githubusercontent.com/20778777/130022984-5db85d98-049c-4a97-922a-c01587278b80.png">
- bold cost names
- lines grey 
- two lines on the top (I remember there was an intention in using the double line)
- more spacing between the lines
- Overheads: Overheads DG (and the other product groups) are actually just DG (Its ok for me to use the short name here)
- Line needs to be above overheads total
- line needs to be over total
- add back the ':' after the price names
- the net price per product group is actually not bold (would also be ok for me to keep it bold)
- Add spacing before the net price for the product groups

For the product group table: can you remove a little of the spacing for the header?  

Another difference is that the product group title appears again with the header for a page break. I assume this cannot be changed and this is not an issue. Just wanted to document the difference here

I can not approve your PR if you want to keep your current solution. I would then wait for @sven1103 to approve your changes or the stakeholders.

jenniferboedker commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy:
```diff
@@ -0,0 +1,370 @@
+        fillDataAnalysisItems(result, content)
+        fillDataManagementItems(result, content)
+        // fill general information
+        fillGeneralInformation(result, content)
```
I still think the name is not a good choice, since the method does not fill "general information". General information in the offer context would be for me: Quotation details, ... stuff from the first page or at least some information or text at all. There are only set some symbols here, so please adjust the naming to that :)

jenniferboedker commented:
We agreed, that we will wait for feedback from the stakeholders, therefore I will give you some feedback on improving the current design you suggested :)
- still think spacing would let it breathe and improves readability (for the cost summary), also add spacing for the lines
- Overheads: (remove the "title", and maybe the keywords
`**Overheads**
  **DG**:                                  50
  **DA**:                                  50
  **PM & DS**:                        50
  **Overheads Total (0%)**:  50`
- why do you list overheads 2x? Its there as overheads total and then as overheads just before VAT
- also I still think bold names for the price values look better.
- remove the final line under total
- make the bold line light grey and not bold

KochTobi commented:
> * why do you list overheads 2x? Its there as overheads total and then as overheads just before VAT
No these are two separate lists. The one sums the overheads the other sums the offer.
> * also I still think bold names for the price values look better.
I don't think so 😄 
> * remove the final line under total
I thought you requested those?

jenniferboedker commented:
> No these are two separate lists. The one sums the overheads the other sums the offer.
Ok, this is really confusing to me. I would not do it like this. But I guess we have very different opinions on that, so I will only focus on implementational feedback.

> I thought you requested those?
I requested to move the line from above total to under total. I think we don't need both.

sven1103 commented:
Ok, technically this is really fantastic, you have to teach me how this works! Great work!

I have made some comments to the design, but please let this not distract you from your achievement here!
[doc00211620210831113021.pdf](https://github.com/qbicsoftware/offer-manager-2-portlet/files/7082839/doc00211620210831113021.pdf)

sven1103 commented:
Ah and I just saw that the item numbers are reset in every group and start again at `1`. It should be continuous.

jenniferboedker commented:
> Ok, technically this is really fantastic, you have to teach me how this works! Great work!
> 
I would also be very interested in knowing how you did it! Maybe you can present it to the team? I am pretty sure at least Steffen would be interested too :)

KochTobi commented:
@sven1103, @jenniferboedker I addressed all your comments. Could you have another look?

sven1103 reviewed (changes requested):
I think you are nearly there:

![image](https://user-images.githubusercontent.com/9976560/132458116-c5213a9c-5f65-4329-ac3d-cbb736f93193.png)

Can you please take measures such that the header image border aligns with the text block.

![image](https://user-images.githubusercontent.com/9976560/132458685-11fa8314-713e-4dad-ab0e-9ad8f97118b1.png)

The overview section still is not in the design it was. Can you please adjust the line color and convert the triple line to double line.

![image](https://user-images.githubusercontent.com/9976560/132458888-16e93737-ee57-4f3f-952e-e9fe59525174.png)

To me it looks like the font size for the description text changed. Also the spacing of the cost overview lines are now to big. When you download an offer created with the current production version, you see immediately what I mean. Maybe we can quickly chat, and adjust it properly.

KochTobi commented:
> I think you are nearly there:
> 
> ![image](https://user-images.githubusercontent.com/9976560/132458116-c5213a9c-5f65-4329-ac3d-cbb736f93193.png)
> 
> Can you please take measures such that the header image border aligns with the text block.

I will happily do that. Do you have a header image without a white border?

> ![image](https://user-images.githubusercontent.com/9976560/132458685-11fa8314-713e-4dad-ab0e-9ad8f97118b1.png)
> 
> The overview section still is not in the design it was. Can you please adjust the line color and convert the triple line to double line.
> 

> ![image](https://user-images.githubusercontent.com/9976560/132458888-16e93737-ee57-4f3f-952e-e9fe59525174.png)
> 
> To me it looks like the font size for the description text changed. Also the spacing of the cost overview lines are now to big. When you download an offer created with the current production version, you see immediately what I mean. Maybe we can quickly chat, and adjust it properly.

Did you generate it locally or with the portal. There are system settings that overwrite the font in the liferay instance.
I increased the spacing upon your request. The original spacing was made with pixels so it depends on the device that requests the offer how large the spacing ist.
Do you have a spacing (mm) in mind?

KochTobi commented:
> To me it looks like the font size for the description text changed. 

What do you mean by that? Do you mean `EXPERIMENTAL DESIGN` ? This is the experimental design of the offer you generated 😄 it was set to `EXPERIMENTAL DESIGN` in the offer.

sven1103 reviewed (approved):
Great now, its perfect imo!
